File: src/test_rom_nestest.cpp
#include "NES.h"
#include <stdint.h>
#include <unordered_map>
#include <vector>
#include <stdio.h>

std::vector<std::unordered_map<std::string, uint64_t>>
_parse_log_entries() {
    std::string log_path = "../roms/nestest/nestest.log";
    FILE* f = fopen(log_path.c_str(), "rb");
    if (NULL == f) {
        printf("Failed to get log entires\n");
        exit(1);
    }

    std::vector<std::unordered_map<std::string, uint64_t>> entries;

    char buffer[255] = { 0 };
    while (fgets((char*)buffer, 255, f)) {
        std::string line = std::string(buffer);
        // printf("%s\n", line.c_str());
        // Most of the values are base 16 except ppu, scanline and cycles.
        uint64_t pc = std::stoull(line.substr(0, 4), 0, 16);
        uint64_t a = std::stoull(line.substr(50, 2), 0, 16);
        uint64_t x = std::stoull(line.substr(55, 2), 0, 16);
        uint64_t y = std::stoull(line.substr(60, 2), 0, 16);
        uint64_t p = std::stoull(line.substr(65, 2), 0, 16);
        uint64_t sp = std::stoull(line.substr(71, 2), 0, 16);
        uint64_t ppu_cycles = std::stoull(line.substr(78, 3));
        uint64_t scanline = std::stoull(line.substr(82, 3));
        uint64_t cpu_cycles = std::stoull(line.substr(90, line.length()));

        std::unordered_map<std::string, uint64_t> entry = {
            { "pc", pc },
            { "a", a },
            { "x", x },
            { "y", y },
            { "p", p },
            { "sp", sp },
            { "ppu_cycles", ppu_cycles },
            { "scanline", scanline },
            { "cpu_cycles", cpu_cycles }
        };
        entries.emplace_back(entry);
    }

    fclose(f);
    return entries;
}

void
print_failure_state(
    std::unordered_map<std::string, uint64_t>& entry,
    NES& nes
) {
    printf(
        "%llx %llx %llx %llx %llx %llx %llu %llu %llu (entry)\n",
        entry["pc"],
        entry["a"],
        entry["x"],
        entry["y"],
        entry["p"],
        entry["sp"],
        entry["cpu_cycles"],
        entry["ppu_cycles"],
        entry["scanline"]
    );
    printf(
        "%x %x %x %x %x %x %llu %llu %u (cpu)\n",
        nes.cpu.pc,
        nes.cpu.a,
        nes.cpu.x,
        nes.cpu.y,
        nes.cpu.p,
        nes.cpu.sp,
        nes.cpu.clock.cpu_cycles,
        nes.cpu.clock.ppu_cycles,
        nes.ppu.scanline
    );
}

int
main() {
    std::string filepath = "../roms/nestest/nestest.nes";
    auto entries = _parse_log_entries();

    NES nes = NES(filepath);
    nes.cpu.pc = 0xc000;

    for (auto& entry : entries) {
        if (nes.cpu.a != entry["a"]) {
            printf("Failed a\n");
            print_failure_state(entry, nes);
            exit(1);
        }
        if (nes.cpu.x != entry["x"]) {
            printf("Failed x\n");
            print_failure_state(entry, nes);
            exit(1);
        }
        if (nes.cpu.y != entry["y"]) {
            printf("Failed y\n");
            print_failure_state(entry, nes);
            exit(1);
        }
        if (nes.cpu.p != entry["p"]) {
            printf("Failed p\n");
            print_failure_state(entry, nes);
            exit(1);
        }
        if (nes.cpu.pc != entry["pc"]) {
            printf("Failed pc\n");
            print_failure_state(entry, nes);
            exit(1);
        }
        if (nes.cpu.sp != entry["sp"]) {
            printf("Failed sp\n");
            print_failure_state(entry, nes);
            exit(1);
        }
        if (nes.cpu.clock.cpu_cycles != entry["cpu_cycles"]) {
            printf("Failed cpu cycles\n");
            print_failure_state(entry, nes);
            exit(1);
        }
        if (nes.ppu.clock.ppu_cycles != entry["ppu_cycles"]) {
            printf("Failed ppu cycles\n");
            print_failure_state(entry, nes);
            exit(1);
        }
        if (nes.ppu.scanline != entry["scanline"]) {
            printf("Failed scanline\n");
            print_failure_state(entry, nes);
            exit(1);
        }
        nes.cpu.next(false);
    }
    return 0;
}

File: py/tests/test_rom_nestest.py
from pathlib import Path

import pytest

from nes import NES


def _parse_log_entry(entry: str) -> dict:
    # TODO
    # For now just grab the address, register states, and cycles
    # for the CPU. In future we can look at instructions, operands
    # and opcode bytes, as well as PPU.
    entry = entry.rstrip('\n')
    pc = int(entry[0:4], base=16)
    a = int(entry[50:52], base=16)
    x = int(entry[55:57], base=16)
    y = int(entry[60:62], base=16)
    p = int(entry[65:67], base=16)
    sp = int(entry[71:73], base=16)
    ppu_cycles = int(entry[78:81])
    scanline = int(entry[82:85])
    cpu_cycles = int(entry[90:len(entry)])

    return {
        'pc': pc,
        'a': a,
        'x': x,
        'y': y,
        'p': p,
        'sp': sp,
        'ppu_cycles': ppu_cycles,
        'scanline': scanline,
        'cpu_cycles': cpu_cycles,
    }

def test_nestest():
    """
    Runs the nestest ROM, comparing the CPU and memory states
    to the nestest.log. This tests the CPU directly without looking
    at the PPU at all.
    """

    rom_path = Path(__file__).parent.joinpath('../roms/nestest/nestest.nes')
    log_path = Path(__file__).parent.joinpath('../roms/nestest/nestest.log')

    with open(log_path) as file:
        log_entries = [_parse_log_entry(line) for _, line in enumerate(file)]

    nes = NES(rom_path)
    # Start here, not where the RESET vector says.
    nes.cpu.pc = 0xc000

    for i, entry in enumerate(log_entries):
        assert nes.cpu.a == entry['a']
        assert nes.cpu.x == entry['x']
        assert nes.cpu.y == entry['y']
        assert nes.cpu.p == entry['p']
        assert nes.cpu.pc == entry['pc']
        assert nes.cpu.sp == entry['sp']
        assert nes.cpu._clock.cpu_cycles == entry['cpu_cycles']
        assert nes.ppu._clock.ppu_cycles == entry['ppu_cycles']
        assert nes.ppu.scanline == entry['scanline']

        # We're always 1 behind cpu.next i believe, so if we run into
        # an error on an entry it's because the previous instruction
        # failed and as a result this entry's state isn't what the
        # last instruction should've caused.
        nes.run()
